File: bench/suite/src/st/k-nucleotide/javascript/k-nucleotide.js
// k-nucleotide: single-threaded, plain JavaScript (Benchmarks Game algorithm:
// extract sequence THREE, count every k-nucleotide with a hash table keyed by
// the k-length substring). Reads stdin, ignores argv.
'use strict';
const fs = require('fs');

function readSequence() {
  const data = fs.readFileSync(0, 'latin1');
  const start = data.indexOf('>THREE');
  const bodyStart = data.indexOf('\n', start) + 1;
  let end = data.indexOf('>', bodyStart);
  if (end < 0) end = data.length;
  return data.slice(bodyStart, end).replace(/\n/g, '').toUpperCase();
}

function frequencies(seq, k) {
  const counts = new Map();
  const last = seq.length - k;
  for (let i = 0; i <= last; i++) {
    const key = seq.substring(i, i + k);
    counts.set(key, (counts.get(key) || 0) + 1);
  }
  return counts;
}

function sortedFrequencies(seq, k) {
  const counts = frequencies(seq, k);
  const total = seq.length - k + 1;
  const entries = Array.from(counts.entries());
  entries.sort((a, b) => (b[1] - a[1]) || (a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : 0));
  let out = '';
  for (const [key, c] of entries) out += key + ' ' + (c * 100 / total).toFixed(3) + '\n';
  return out + '\n';
}

function specificCount(seq, fragment) {
  const counts = frequencies(seq, fragment.length);
  return (counts.get(fragment) || 0) + '\t' + fragment + '\n';
}

const seq = readSequence();
let out = sortedFrequencies(seq, 1) + sortedFrequencies(seq, 2);
for (const f of ['GGT', 'GGTA', 'GGTATT', 'GGTATTTTAATT', 'GGTATTTTAATTTATAGT']) {
  out += specificCount(seq, f);
}
process.stdout.write(out);

File: bench/suite/src/st/k-nucleotide/python/k-nucleotide.py
# k-nucleotide: single-threaded, pure CPython (Benchmarks Game algorithm:
# extract sequence THREE, count every k-nucleotide with a dict keyed by the
# k-length substring). Reads stdin, ignores argv.
import sys


def read_sequence():
    data = sys.stdin.buffer.read()
    start = data.index(b">THREE")
    body_start = data.index(b"\n", start) + 1
    end = data.find(b">", body_start)
    if end < 0:
        end = len(data)
    return data[body_start:end].replace(b"\n", b"").upper().decode("latin-1")


def frequencies(seq, k):
    counts = {}
    get = counts.get
    for i in range(len(seq) - k + 1):
        key = seq[i:i + k]
        counts[key] = get(key, 0) + 1
    return counts


def sorted_frequencies(seq, k):
    counts = frequencies(seq, k)
    total = len(seq) - k + 1
    items = sorted(counts.items(), key=lambda kv: (-kv[1], kv[0]))
    return "".join("%s %.3f\n" % (key, c * 100.0 / total) for key, c in items) + "\n"


def main():
    seq = read_sequence()
    out = [sorted_frequencies(seq, 1), sorted_frequencies(seq, 2)]
    for frag in ("GGT", "GGTA", "GGTATT", "GGTATTTTAATT", "GGTATTTTAATTTATAGT"):
        out.append("%d\t%s\n" % (frequencies(seq, len(frag)).get(frag, 0), frag))
    sys.stdout.write("".join(out))


main()
